src/vtl.ts

```diff
@@ -418,4 +437,78 @@ export class VTL {
     const __exhaustive: never = node;
     return __exhaustive;
   }
+
+  private mapForEachArgs(call: CallExpr) {
+    const fn = call.args.callbackfn as FunctionExpr;
+    return fn.parameters.map((p) => p.name);
+  }
+
+  private renderMapOrForEach(
+    call: CallExpr,
+    list: string,
+    // Should start with $
+    returnVariable?: string,
+    index?: string,
+    array?: string
+  ) {
+    if (index) {
+      this.add(`#set($${index} = $foreach.index)`);
+    }
+    if (array) {
+      this.add(`#set($${array} = ${list})`);
+    }
+
+    const fn = call.args.callbackfn as FunctionExpr;
+
+    const tmp = returnVariable ? returnVariable : this.newLocalVarName();
+
+    for (const stmt of fn.body.statements) {
+      this.eval(stmt, tmp);
+    }
+
+    return tmp;
+  }
+
+  /**
+   * Recursively dives into map operations until a non-map or a map with arr paremeter is found.
+   * evaluates the expression after the last map.
+   *
+   * @return [firstVariable, list variable, render function]
+   */
+  private divePropAccessMapOperations(
+    expr: Expr,
+    // Should start with $
+    returnVariable: string,
+    alwaysEvaluate?: boolean
+  ): [string, string, () => void] {
+    if (
+      !alwaysEvaluate &&
+      expr.kind === "CallExpr" &&
+      expr.expr.kind === "PropAccessExpr" &&
+      expr.expr.name === "map"
+    ) {
+      const [value, index, array] = this.mapForEachArgs(expr);
+
+      const next = expr.expr.expr;
+
+      // If we find array, stop recursing as array wouldn't be correct in the current map.
+      const [lastValue, list, lastRender] = this.divePropAccessMapOperations(
+        next,
+        `$${value}`,
+        !!array
+      );
+
+      return [
+        lastValue,
+        list,
+        () => {
+          lastRender();
+
+          this.renderMapOrForEach(expr, list, returnVariable, index, array);
+        },
+      ];
+    }
+    // If the expression isn't a map, return the expression and return variable, render nothing
+    return [returnVariable, this.eval(expr), () => {}];
```

Comment:
`() => {}`?

Reply:
The base case shouldn't do anything, can change it to be undefined.

Reply:
This smells a bit. Makes me think the callback is unnecessary. Base case should be a return from a recursive call instead of an empty callback?

Reply:
The "callbacks" give the caller control of when the computed content is applied to the mutable state. Its not a callback, but a renderer trigger. 

Another option is to pass a callback that contains the "before" values (the foreach and other things) into this method, but I like that less because its more of a callback pattern than the late rendering closure. 

And the other option is to build an array of VTL statements instead of mutating the cental statements array, then this wouldn't be an issue because the computed statements could be returned and injected in where needed.

Reply:
I thought the recursive function would recurse until it finds the terminal case. Terminal case would start the `#foreach` and return a string pointing to the result of its computation. The recursive cases would then emit their statements pointing at that.

Reply:
I'll give it a try.

I was trying to keep the method independent of the calling implementation, where the foreach is part of the parent, not the flattened maps. 

Looking at the outcome, it appears the only part that comes before the render is the foreach, so maybe it's fine. It'll be a bit strange because the foreach will be in the method, but the end will be outside.